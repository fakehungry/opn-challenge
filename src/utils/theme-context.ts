import { Dispatch, SetStateAction, createContext } from 'react';
import lightTheme from '../styles/light-theme';

export enum ThemeMode {
  LIGHT = 'light',
  DARK = 'dark',
}

type ThemeContextType = {
  themeMode: ThemeMode;
  theme: typeof lightTheme;
  setThemeMode: Dispatch<SetStateAction<ThemeMode | null>>;
};

export const ThemeContext = createContext<ThemeContextType>({
  themeMode: ThemeMode.LIGHT,
  theme: lightTheme,
  setThemeMode: () => {},
});
