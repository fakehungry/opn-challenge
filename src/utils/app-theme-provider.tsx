import { useEffect, useState } from 'react';
import { ThemeContext, ThemeMode } from './theme-context';
import lightTheme from '../styles/light-theme';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/styled-component-theme';
import darkTheme from '../styles/dark-theme';

type AppThemeProviderProps = {
  children: JSX.Element | JSX.Element[];
  theme?: ThemeMode;
};

const getThemeFromLocalStorage = () =>
  window.localStorage.getItem('theme') || ThemeMode.LIGHT;

const setThemeToLocalStorage = (theme: ThemeMode) =>
  window.localStorage.setItem('theme', theme);

export const AppThemeProvider = (props: AppThemeProviderProps) => {
  const { children, theme: themeProp } = props;
  const [themeMode, setThemeMode] = useState<ThemeMode | null>(null);
  const colors = themeMode === ThemeMode.LIGHT ? lightTheme : darkTheme;

  useEffect(() => {
    if (themeProp) {
      setThemeMode(themeProp);
    } else if (themeMode === null) {
      setThemeMode(getThemeFromLocalStorage() as ThemeMode);
      return;
    }

    setThemeToLocalStorage(themeMode as ThemeMode);
  }, [themeMode, themeProp]);

  return (
    <ThemeContext.Provider
      value={{ themeMode: themeMode as ThemeMode, theme: colors, setThemeMode }}
    >
      <ThemeProvider theme={{ ...theme, colors: colors }}>
        {children}
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
