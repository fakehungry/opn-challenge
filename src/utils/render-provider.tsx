import React from 'react';
import { render } from '@testing-library/react';
import { AppThemeProvider } from './app-theme-provider';
import { ThemeMode } from './theme-context';

// REMARK: Test only light mode
export const renderProvider = (props: any) => {
  return render(
    <AppThemeProvider theme={ThemeMode.LIGHT}>{props}</AppThemeProvider>,
  );
};
