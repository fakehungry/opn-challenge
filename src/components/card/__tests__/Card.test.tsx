import { fireEvent } from '@testing-library/react';
import Card from '..';
import { renderProvider } from '../../../utils/render-provider';

describe('card component', () => {
  const props = {
    place: 'Baan Kru Noi',
    imgsrc: '/images/baan-kru-noi.jpg',
    imgalt: 'Baan Kru Noi',
  };

  it('should render card component and match snapshot', () => {
    const { container } = renderProvider(<Card {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render card component with overlay and match snapshot', () => {
    const { container, getByText } = renderProvider(<Card {...props} />);
    const donateBtn = getByText('Donate');
    fireEvent.click(donateBtn);
    expect(container).toMatchSnapshot();
  });

  it('should close overlay when X mark icon is clicked', () => {
    const { getByText, queryByText, getByTestId } = renderProvider(
      <Card {...props} />,
    );
    const donateBtn = getByText('Donate');
    fireEvent.click(donateBtn);
    fireEvent.click(getByTestId('x-mark-icon') as HTMLElement);
    expect(queryByText('Select the amount to donate (THB)')).toBeNull();
  });

  it('should close overlay when pay btn is clicked', () => {
    const { getByText, queryByText } = renderProvider(<Card {...props} />);
    const donateBtn = getByText('Donate');
    fireEvent.click(donateBtn);
    fireEvent.click(getByText('Pay') as HTMLElement);
    expect(queryByText('Select the amount to donate (THB)')).toBeNull();
  });

  it('should render card component with selected amount and match snapshot', () => {
    const { container, getByText } = renderProvider(<Card {...props} />);
    const donateBtn = getByText('Donate');
    fireEvent.click(donateBtn);
    fireEvent.click(getByText('100') as HTMLElement);
    expect(container).toMatchSnapshot();
  });
});
