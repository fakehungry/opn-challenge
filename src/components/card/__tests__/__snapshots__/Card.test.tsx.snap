// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card component should render card component and match snapshot 1`] = `
<div>
  <div
    class="sc-beySPh csDOYD"
    imgalt="Baan Kru Noi"
    imgsrc="/images/baan-kru-noi.jpg"
    place="Baan Kru Noi"
  >
    <img
      alt="Baan Kru Noi"
      src="/images/baan-kru-noi.jpg"
    />
    <div
      class="row"
    >
      <h2>
        Baan Kru Noi
      </h2>
      <button
        class="sc-guDLey hsxNFn"
      >
        Donate
      </button>
    </div>
  </div>
</div>
`;

exports[`card component should render card component with overlay and match snapshot 1`] = `
<div>
  <div
    class="sc-beySPh csDOYD"
    imgalt="Baan Kru Noi"
    imgsrc="/images/baan-kru-noi.jpg"
    place="Baan Kru Noi"
  >
    <img
      alt="Baan Kru Noi"
      src="/images/baan-kru-noi.jpg"
    />
    <div
      class="row"
    >
      <h2>
        Baan Kru Noi
      </h2>
      <button
        class="sc-guDLey hsxNFn"
      >
        Donate
      </button>
    </div>
    <div
      class="overlay overlay-anim"
    >
      <svg
        data-testid="x-mark-icon"
        style="width: 20px; height: 20px;"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
          fill="#5e6778"
        />
      </svg>
      <div
        class="column"
      >
        <h2>
          Select the amount to donate (THB)
        </h2>
        <div
          class="radio-group"
        >
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-10"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-10"
              name="donate-Baan Kru Noi"
              type="radio"
              value="10"
            />
            <span>
              10
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-20"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-20"
              name="donate-Baan Kru Noi"
              type="radio"
              value="20"
            />
            <span>
              20
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-50"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-50"
              name="donate-Baan Kru Noi"
              type="radio"
              value="50"
            />
            <span>
              50
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-100"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-100"
              name="donate-Baan Kru Noi"
              type="radio"
              value="100"
            />
            <span>
              100
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-500"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-500"
              name="donate-Baan Kru Noi"
              type="radio"
              value="500"
            />
            <span>
              500
            </span>
          </label>
        </div>
        <button
          class="sc-guDLey hsxNFn"
        >
          Pay
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`card component should render card component with selected amount and match snapshot 1`] = `
<div>
  <div
    class="sc-beySPh csDOYD"
    imgalt="Baan Kru Noi"
    imgsrc="/images/baan-kru-noi.jpg"
    place="Baan Kru Noi"
  >
    <img
      alt="Baan Kru Noi"
      src="/images/baan-kru-noi.jpg"
    />
    <div
      class="row"
    >
      <h2>
        Baan Kru Noi
      </h2>
      <button
        class="sc-guDLey hsxNFn"
      >
        Donate
      </button>
    </div>
    <div
      class="overlay overlay-anim"
    >
      <svg
        data-testid="x-mark-icon"
        style="width: 20px; height: 20px;"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
          fill="#5e6778"
        />
      </svg>
      <div
        class="column"
      >
        <h2>
          Select the amount to donate (THB)
        </h2>
        <div
          class="radio-group"
        >
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-10"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-10"
              name="donate-Baan Kru Noi"
              type="radio"
              value="10"
            />
            <span>
              10
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-20"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-20"
              name="donate-Baan Kru Noi"
              type="radio"
              value="20"
            />
            <span>
              20
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-50"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-50"
              name="donate-Baan Kru Noi"
              type="radio"
              value="50"
            />
            <span>
              50
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-100"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-100"
              name="donate-Baan Kru Noi"
              type="radio"
              value="100"
            />
            <span>
              100
            </span>
          </label>
          <label
            class="sc-hLQSwg ixbFbr"
            for="donate-Baan Kru Noi-500"
          >
            <input
              class="sc-dmyCSP gYNgzD"
              id="donate-Baan Kru Noi-500"
              name="donate-Baan Kru Noi"
              type="radio"
              value="500"
            />
            <span>
              500
            </span>
          </label>
        </div>
        <button
          class="sc-guDLey hsxNFn"
        >
          Pay
        </button>
      </div>
    </div>
  </div>
</div>
`;
